src/components/DownloadButton.js

```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import { Button } from 'antd';
+
+const META_TYPE = 'text/json';
+
+function ExportButton({ ast }) {
+  const data = JSON.stringify(ast, null, 2);
```

Comment:
Instead of expensively creating the href on every ast change, which the user might not click the "Download" button, we can create the `<a>` tag with the `href` only on click:

```jsx
const onClick = () => {
    const a = document.createElement('a');
    a.href = createHref(ast);
    a.download = 'data.json';
    a.click();
  }

  return (
    <Button onClick={onClick}>Download AST</Button>
  );
```

this way, you don't need to memo too.

Reply:
@tanhauhau Fixed, pls check